pkg/operator/admission/init.go

```diff
@@ -0,0 +1,175 @@
+package admission
+
+import (
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"os"
+)
+
+const (
+	containerName      = "rook-ceph-operator"
+	secretName         = "rook-admission-controller-secret"
+	acServiceName      = "rook-admission-controller-service"
+	webhookConfigName  = "rook-ceph-webhook"
+	kubeNamespace      = "kube-system"
+	configMapName      = "extension-apiserver-authentication"
+	secretVolumeName   = "webhook-certs"
+	secretMountPath    = "/etc/webhook"
+	serviceAccountName = "rook-ceph-system"
+	storageFlag        = "ceph"
+)
+
+var (
+	validatePath       = "/validate"
+	servicePort  int32 = 443
+	// TODO Make it configurable
+	serverPort int32 = 8079
+	namespace        = os.Getenv(k8sutil.PodNamespaceEnvVar)
+)
+
+func isSecretPresent(context *clusterd.Context) (bool, error) {
+
+	_, err := context.Clientset.CoreV1().Secrets(namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		// If secret is not found. All good ! Proceed with rook without admission controllers
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func createWebhookService(context *clusterd.Context) error {
+
+	webhookService := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      acServiceName,
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Port: servicePort,
+					TargetPort: intstr.IntOrString{
+						IntVal: serverPort,
+					},
+				},
+			},
+
+			Selector: map[string]string{
+				"app": "rook-ceph-admission-controller",
```

Comment:
I would expect everything in `pkg/operator/admission` to be independent from the storage provider. If there is something specific to ceph it should go under `pkg/operator/ceph/admission`. This separation will make it possible for other storage providers to implement the admission controller.

Reply:
To implement, you may want to define an interface that the ceph admission controller would be the first to implement.